File: CrawlManager/models/rabbit_client.rb
class RabbitClient
  def initialize
    unless ENV["APP_ENV"]=="test"
      timeout = ENV["APP_ENV"]=="production" ? 1 : 100
      conn = Bunny.new(hostname: ENV["RABBITMQ_HOST"], continuation_timeout: timeout)
      conn.start
      @ch = conn.create_channel
      #@ch.prefetch(100)
    end
  end
  
  def send(msg:, queue_name:)
    unless ENV["APP_ENV"]=="test"
      if msg.nil?
        raise StandardError("tried to send a nil message to RabbitMQ")
      end
      q = @ch.queue(queue_name, :durable => true)
      q.publish(msg.to_json, :persistent => true, :content_type => "application/json")
    end
  end
  
  def listen(queue_name:)
    unless ENV["APP_ENV"]=="test"
      q = @ch.queue(queue_name, :durable => true)
      q.subscribe(:manual_ack => true, :block => true) do |delivery_info, properties, body|
        yield body
        @ch.ack(delivery_info.delivery_tag)
      end
    end
  end
  
  def purge(queue_name:)
    unless ENV["APP_ENV"]=="test"
      q = @ch.queue(queue_name, :durable => true)
      q.purge
    end
  end
end

File: PageDownloader/workspace/src/github.com/vdaubry/page_downloader/rabbit_client.go
package main

import (
  "fmt"
  "github.com/streadway/amqp"
  "log"
  "encoding/json"
)

type RabbitClient struct {
  ch *amqp.Channel
}

func failOnError(err error, msg string) {
  if err != nil {
    log.Fatalf("%s: %s", msg, err)
    panic(fmt.Sprintf("%s: %s", msg, err))
  }
}

func connectRabbit(connectionString string) *amqp.Channel {
  conn, err := amqp.Dial(connectionString)
  failOnError(err, "Failed to connect to RabbitMQ")
  
  ch, err := conn.Channel()
  failOnError(err, "Failed to open a channel")

  err = ch.Qos(
    20,     // prefetch count
    0,     // prefetch size
    false, // global
  )
  failOnError(err, "Failed to set QoS")
  
  return ch
}

func (client RabbitClient) queue(queueName string) amqp.Queue {
  q, err := client.ch.QueueDeclare(
    queueName,    // name
    true,         // durable
    false,        // delete when unused
    false,        // exclusive
    false,        // no-wait
    nil,          // arguments
  )
  failOnError(err, "Failed to declare a queue")
  
  return q
}

func NewRabbitClient(connectionString string) *RabbitClient {
  ch := connectRabbit(connectionString)
  return &RabbitClient{ch}
}

func (client RabbitClient) Listen(queueName string) {
  q := client.queue(queueName)
  
  msgs, err := client.ch.Consume(
    q.Name, // queue
    "",     // consumer
    false,  // auto-ack
    false,  // exclusive
    false,  // no-local
    false,  // no-wait
    nil,    // args
  )
  failOnError(err, "Failed to register a consumer")
  
  
  forever := make(chan bool)
  responses := make(chan *HttpResponse)
  
  go func() {
    for message := range msgs {
      webPage := &WebPage{}
      json.Unmarshal([]byte(message.Body), &webPage)
      log.Printf("Received url to download: %s", webPage.Url)
      
      go webPage.Download(responses)
      select {
      case response := <-responses:
        if response.err == nil {
          fmt.Printf("%status: %s in %s\n",
                 response.response.Status, response.totalTime)
        } else {
          fmt.Printf("err: %s in %s\n", response.err, response.totalTime)
        }
        message.Ack(false)
      default:
      }
      
    }
  }()

  log.Printf("Waiting for messages...")
  <-forever
}

func (client RabbitClient) Send(queueName string, message []byte) {
  q := client.queue(queueName)
  
  err := client.ch.Publish(
    "",           // exchange
    q.Name,       // routing key
    false,        // mandatory
    false,
    amqp.Publishing{
      DeliveryMode: amqp.Persistent,
      ContentType:  "text/plain",
      Body:         message,
    })
  failOnError(err, "Failed to publish a message")
}